### PR TITLE
bpo-37146: Allow to configure opcode cache and fix huntleaks by presetting it to 1

### PR DESCRIPTION
**Pull request by pablogsal**
https://bugs.python.org/issue37146

### CODE REVIEW

File: Include/internal/pycore_pystate.h [hunk 1]
```diff
@@ -43,6 +43,10 @@ struct _pending_calls {
     int last;
 };
 
+struct _opcodecache_config {
+    int minruns;
+};
```
Comment (vstinner) [inline Include/internal/pycore_pystate.h]: the struct is useless, just put directly the int into _ceval_runtime_state.

**Comment by pablogsal** [comment]
Tested locally:

```
❯ ./python -m test -R 3:3 test_pprint
Run tests sequentially
0:00:00 load avg: 1.57 [1/1] test_pprint
beginning 6 repetitions
123456
......

== Tests result: SUCCESS ==

1 test OK.

Total duration: 751 ms
Tests result: SUCCESS
```

File: Include/internal/pycore_pystate.h [hunk 2]
```diff
@@ -312,6 +317,10 @@ PyAPI_FUNC(void) _PyInterpreterState_DeleteExceptMain(_PyRuntimeState *runtime);
 
 PyAPI_FUNC(void) _PyGILState_Reinit(_PyRuntimeState *runtime);
 
+PyAPI_FUNC(void) _PyEval_SetOpcodeCacheMinRuns(int minruns);
+
+PyAPI_FUNC(int) _PyEval_GetOpcodeCacheMinRuns();
```
Comment (vstinner) [inline Include/internal/pycore_pystate.h]: you should use: (void)

File: Lib/test/libregrtest/runtest.py [hunk 3]
```diff
@@ -206,6 +206,8 @@ def _test_module(the_module):
 def _runtest_inner2(ns, test_name):
     # Load the test function, run the test function, handle huntrleaks
     # and findleaks to detect leaks
+    minruns = sys._getopcacheminruns()
```
Comment (vstinner) [inline Lib/test/libregrtest/runtest.py]: I don't think that it matters to restore the old value. I'm not sure that sys._getopcacheminruns() is needed.

File: Python/pylifecycle.c [hunk 4]
```diff
@@ -737,6 +737,9 @@ _Py_PreInitializeFromPyArgv(const PyPreConfig *src_config, const _PyArgv *args)
         return status;
     }
 
+    // XXX: Add this properly to the commandline config
```
Comment (vstinner) [inline Python/pylifecycle.c]: pleave remove this XXX. It doesn't *have to* be a cmdline option.

**Comment by pablogsal** [comment]
Tested also with `./python -m test -R 3:3 -j0 -r -x test_asyncio -x test_multiprocessing_spawn`

File: Lib/test/libregrtest/runtest.py [hunk 5]
```diff
@@ -206,6 +206,8 @@ def _test_module(the_module):
...
+    sys._setopcacheminruns(1)
```
Comment (vstinner) [inline Lib/test/libregrtest/runtest.py]: That's wrong. Only dash_R() should be modified in libregrtest/refleak.py. You must not disable the opcache in runtest.py.

**Review by vstinner — CHANGES_REQUESTED** [review]

**Comment by methane** [comment]
The code object will be optimized only when `++co->co_opcache_flag == opcacheminruns`.
So decreasing min_runs by `_setopcacheminruns()`will cause some hot codes will be not
optimized forever.  I don't want to expose such switch.

Please use #13787 for 3.8b1.

**Comment by vstinner** [comment]
> Please use #13787 for 3.8b1.

I like #13787, it's a reasonable workaround for beta1. We can design a better fix after the beta.

File: Lib/test/libregrtest/runtest.py (hunk 5 above)
Comment (pablogsal) [reply Lib/test/libregrtest/runtest.py]: Sorry, this was supposed to reset the original value. Check out my last fixup (bd3a951ff81eab28b77d05290a2466573c41eda2),

File: Include/internal/pycore_pystate.h (hunk 1 above)
Comment (pablogsal) [reply Include/internal/pycore_pystate.h]: I made a struct for easy extending more stuff in the config in the future.

I will remove it

File: Lib/test/libregrtest/runtest.py (hunk 5 above)
Comment (vstinner) [reply Lib/test/libregrtest/runtest.py]: I think you misunderstood my comment. You must not modify this function but dash_R() function.

**Comment by pablogsal** [comment]
Ok, I will close this PR then.

**Comment by vstinner** [comment]
> Ok, I will close this PR then.

Maybe keep it around, it's sounds like a good fix for post-beta1.

**Pull request closed without merging**